### ISSUE TITLE
Right clicking icon and choosing 'Quit' just minimizes

### ISSUE THREAD

**Issue by DivineOmega** [mattermost/desktop#90]
Running Ubuntu 14.04 and Mattermost Desktop (electron-mattermost) v1.0.7, steps to reproduce:
1. Start up Mattermost Desktop in the normal manner.
2. Right click the icon representing the running program in Unity.
3. Click the 'Quit' menu option.

Expected behaviour:
- Mattermost Desktop exits.

Actual behaviour:
- Mattermost Desktop minimizes if not already minimized and does nothing if it is already minimized.

The only way to quit Mattermost Desktop in this instance is to use the File -> Quit menu option or to kill the process.

**Comment by yuya-oc** [comment]
On Linux, the application performs 'minimize' instead of 'close' in order to keep the window. This feature is working in this issue. Unfortunately, I don't know what is the difference between Unity's 'Quit' and normal 'close'.

**Comment by brumle80** [comment]
same "issue" in Fedora 23, click the x and it minimizes, but doesn't close.

**Comment by yuya-oc** [comment]
@brumle80 Did you mean about window's close button? If you did, the 'x' behavior is by design as I had written at the previous post. I think that the point of this issue is the handling for 'Quit' operation caused by desktop environment like Unity/GNOME.

**Comment by jasonblais** [comment]
Thanks @DivineOmega and @brumle80 for the feedback,

A related ticket #178 has been resolved and closed, which should fix the issue presented here.

However, if there is a further improvement you would like to see, I invite you to help contribute your suggestion in Mattermost [feature idea forum](https://mattermost.uservoice.com/forums/306457-general) so it can be discussed, upvoted and considered for a ticket accepting pull requests?

If you do add it to the [feature idea forum](https://mattermost.uservoice.com/forums/306457-general), please include a link back to this GitHub Issue. If you're interested in implementing, please say so and we'll prioritize the review.

**Comment by jasonblais** [comment]
I realized there actually is still an issue with right clicking icon and choosing 'Quit'. I have added it as a known issue [to the desktop changelog](https://github.com/mattermost/desktop/pull/206)

**Comment by jasonblais** [comment]
@RaphaelKimmig I think that's fair

**Comment by amyblais** [comment]
Ticket for tracking: https://mattermost.atlassian.net/browse/MM-14130

**Comment by echosa** [comment]
I'm not sure if this is the right issue to tack on to, or if I should create a new one (happy to do so!), but there's an inconsistency in the way "Quit" menu items work in Mattermost, at least in GNOME.

There are are a few "Quit" options but not all work the same:

- In the application window, `File -> Quit`. This actually quits the program.
- From the tray icon, `<right-click> -> Quit`. This actually quits the program.
- From the upper-left application menu (next to `Activities`). `<Application menu> -> Quit`. **This only closes the window/minimizes to the tray, leaving Mattermost open.
- The Close/`X` button in the upper-right of the windows. This also only closes the windows/minimizes to the tray, but that's expected... sort of. See below.

So, of these four options, the third seems wrong. `<Application menu> -> Quit` should completely quit the program, in my opinion. 

Now, I can understand the fourth option, since you're "closing" rather than "quitting". However, closing the window is keeping Mattermost open/in the tray, even though I have the "Leave app running [...]" setting unchecked/turned off. See screenshot. Is this a bug? Shouldn't Mattermost be quitting when I close the window since I have that setting turned off?

![image](https://user-images.githubusercontent.com/133786/55900351-97b0b480-5b8c-11e9-9526-5d7c64146e40.png)

**Comment by bochecha** [comment]
In addition to @echosa's last post, there is a 5th way to quit the app in GNOME:

* in the Activities overview, click the (x) button in the top-right corner of the window thumbnail. Nothing happens at all.

I too have the "Leave app running [...]" setting unchecked/turned off" and would expect all 5 ways to completely quit the program.

Enabling the option should really only minimize the app for the window close button (4th case from @echosa's comment), all other cases should always quit the program no matter what.

**Comment by aunwin** [comment]
Not being able to quit the Desktop Client in the usual way (e.g. klicking the red x button at the top) might cause people to move away from using mattermost.
Under Ubuntu 18_04 LTS I have to use CTRL+q to actually quit the application.

**Comment by supernintendo** [comment]
Hello. I am not a mattermost user but stumbled upon this issue while debugging an issue with the official Slack client (on Fedora 30, 3.32.2).

It appears that there is simply a lack of commonly agreed upon rules when it comes to the UX for closing and "soft closing" applications in the GNU/Linux ecosystem. This issue doesn't really have a defined set of requirements that can be treated as a source of truth. AFAIK the most popular desktop environment, GNOME, does not address this specific UX pattern in its [Human Interface Guidelines](https://developer.gnome.org/hig/stable/). In fact, the only DE I was able to find that had docs mentioning this was [Elementary OS](https://elementary.io/docs/human-interface-guidelines#human-interface-guidelines).

I've been touch testing a myriad of applications and have found a wide variance of behaviours. For example, here's how the official Slack client behaves:

1. Clicking the X in the window title bar _minimizes_ the application.
2. Right clicking on the window title bar and selecting "Close" _minimizes_ the application.
3. Expanding the "Slack" menu in the GNOME title bar at the top of the screen and selecting "Quit" _minimizes_ the application.
4. Using the "Ctrl+Q" keyboard shortcut while Slack is active is the only way to actually *quit* the application.

Compare this to another Electron-based application, Visual Studio Code. It terminates completely in all of these scenarios. In fact, I haven't found a way using GNOME to close all Visual Studio Code windows without terminating the entire process and having to deal with Electron's slow initialization routine the next time I open it.

I think the issue here has more to do with general GNU/Linux usability. There are no universal human interface guidelines across desktop environments. In macOS, this UX pattern works the same in 99% of applications as described in [Apple's HIG](https://developer.apple.com/design/human-interface-guidelines/):

1. The close button is used to close the window, not the application. Running applications are represented in the macOS dock using a dot under the application icon.
2. Running applications are terminated by selecting "Quit" in the application menu. A "Force Quit" option is conditionally displayed depending on the state of the application.
3. Minimized windows always display as individual objects in the dock.

Unlike Apple, the GNU/Linux landscape has no shared human interface guidelines. GNOME will behave differently than KDE than i3 etc... and this is a good thing! Freedom, customization, and all that. But it does lead to a sort of schizophrenic ecosystem of developer and user configurations where nobody can agree so everyone just accepts that every application will have it own way of doing things and if users want a cohesive UX it's up to them to spend the time installing packages and fine tuning their environment.

I'll go ahead and end my rant here by throwing out a plea. __We need a public API / spec for common UX interactions that all desktop environments can implement.__ It's fine if GNOME and KDE want to treat the "close but actually minimize" behaviour differently but it shouldn't be up to individual software projects to handle those conditions. I should be able to pass the same message to _all_ desktop environments and expect that the handling of that message is consistent with the UI framework's internal UX. __It should not be up to the application to deal with UX concerns of the desktop environment.__

tldr; This problem is beyond the scope of mattermost. UX patterns need a common protocol / interface like windowing systems, drivers, etc. and it should be attractive enough to appeal to the larger developer community.

**Comment by skewty** [comment]
Hey look, another >3 year old bug given low importance due to minuscule Linux market share.  

> Ticket for tracking: https://mattermost.atlassian.net/browse/MM-14130

Doesn't have any useful information (also showing lack of care about this issue).

Is this an upstream issue?  I don't have this issue with other Electron apps.  Is it an upstream one?  Does anyone have links to the upstream issue(s)?

**Comment by amyblais** [comment]
Are you still experiencing this issue on desktop v4.4.0?

**Comment by aunwin** [comment]
Yes. Both for mousclick on the close-icon as well as for alt+F4 Hotkey attempts.

**Comment by cookiengineer** [comment]
I wanted to confirm that this is still the case with GNOME 3.36 and later. Not a Unity setup, up-to-date Arch + GNOME stable.

Why doesn't mattermost instead use the appindicator API to land in systray if that's the wanted behaviour - and allow users to disable systray integration in order to close it when they want it?

Currently, users have to manually `pkill` the process, which I think is pretty annoying.

**Comment by raphaelmro** [comment]
I'm having the same experience on Pop!_OS 19.10 x86_64 and GNOME 3.34.3.
The close-icon minimizes the app and the quit button while on the dock is not triggering any close/kill action.

**Comment by cfcohen** [comment]
I'm running Mattermost in KDE.  In my opinion, closing the window (with the X in the top right corner of the window that normally quits an application) should actually quit the application.  This is especially true if the "leave app running in the notification area" is not checked in the settings.  That setting is completely disabled on my system, probably because there's no actual KDE notification support.   The current behavior just feels like the app is intentionally defeating my attempts to quit, including several other contexts in which an app would normally exit, but instead Mattermost just minimizes (even when it's already minimized, resulting _no_ action).  I would recommend just disabling the non-standard close behavior when the setting is disabled.

**Comment by SemaiCZE** [comment]
> Did you mean about window's "close button"? If you did, the behavior is by design as I had written at the previous post. I think that the point of this issue is the handling for 'Quit' operation caused by desktop environment like Unity/GNOME.

@yuya-oc For me it's bad design choice. I'm using Gnome on Fedora and I have UNchecked the _Leave app running in notification area_ switch in settings. On Gnome there is no such thing as notification area and minimizing applications is also not much supported. I'd love to quit Mattermost client the same way as any other app (including Slack). Please reevaluate your design choise. Thanks

**Comment by volca02** [comment]
On version 4.4.1, while using i3 WM on linux, clicking the red X, or using the i3's close application key combo (which seems to use the WM_DELETE protocol) does nothing at all. Regardless whether the option `Leave app running in notification area when application window is closed` is enabled or not.

**Comment by cookiengineer** [comment]
I created the pull request #1279 and it respects the `minimizeToTray` setting correctly.

- If minimizeToTray is set to true, hide the main window.
- If minimizeToTray is set to false, close the main window.

**Comment by dontlaugh** [comment]
Let us quit the app, guys

**Comment by skewty** [comment]
@cookiengineer has even done a PR for the fix.  Is the PR not acceptable?  What is holding this up?

**Comment by cookiengineer** [comment]
@skewty The current status is that I'm building my own fork because nobody answered on the UI/UX implications for this change and I cannot answer these questions, as I am not affiliated with mattermost anyhow.

The settings UI is currently differently rendered on each platform whereas Windows doesn't even have the option, and therefore is behaving wrong in my opinion; and MacOS has this setting but needs testing from a UX member of mattermost.

PS: As an outside contributor I have no clue on what or whether there are even test scenarios for this, and to be honest it seems as the workflow for accepting (or not) pull requests for UX changes seems to be totally broken and overengineered from the CI perspective.

Therefore I kind of gave up and decided to keep maintaining my fork until somebody else has enough time to deal with the UX debate behind this fix. My fork works and behaves the same (given the settings are rendered everywhere) on MacOS, Windows and Linux.

As long as nobody from mattermost's UX team answers, this will likely not be merged in upstream.

**Comment by dontlaugh** [comment]
Thanks for your work so far on this. I'm not affiliated with Mattermost, either. But, as a GNOME desktop user, this is a major quality of life issue for me.

**Comment by plaperdr** [comment]
The problem is still present on Mattermost 4.6.1 on GNOME 3.36.3. Isn't it possible to provide a fix for the Linux version in the meantime so that clicking on the X closes the window (or minimizes it depending on settings)?

**Comment by devinbinnie** [comment]
Closing as we have other issues for this and a toggle checkbox to determine whether this should happen or not.
